endpoint.go

```diff
@@ -235,11 +332,13 @@ func (ep *endpoint) Join(containerID string, options ...EndpointOption) error {
 
 	ep.Unlock()
 	defer func() {
-		ep.Lock()
 		if err != nil {
+			ep.Lock()
 			ep.container = nil
+			ep.Unlock()
+		} else {
+			ep.network.ctrlr.addEndpointToStore(ep)
```

Comment:
I don't think adding to data store operation should be handled in defer function. For starters we are not checking if there was an error while adding to store and if there was what are we going to do about that.